core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala

```diff
@@ -113,11 +114,21 @@ private[spark] class ExternalSorter[K, V, C](
     if (shouldPartition) partitioner.get.getPartition(key) else 0
   }
 
+  private val metaInitialRecords = conf.getInt("spark.shuffle.sort.metaInitialRecords", 256)
+  private val kvChunkSize = conf.getInt("spark.shuffle.sort.kvChunkSize", 8192)
+  private val useSerializedPairBuffer =
+    !ordering.isDefined && conf.getBoolean("spark.shuffle.sort.serializeMapOutputs", true) &&
+    conf.get("spark.serializer", null) == classOf[KryoSerializer].getName
```

Comment:
We should probably check here that there is not a user kryo registrator that has set autoReset to false. Otherwise, the whole thing will not work. It's tricky because Kryo does not expose a getter for the autoReset field. One idea would be to coerce that autoReset is true when we instantiate Kryo... but that could regress behavior for users who were disabling it as an optimization. Maybe we should just make a "best effort" attempt to use reflection to access the `autoReset` field of the `Kryo` object and throw an exception if it is set to `false` and `spark.shuffle.sort.serializeMapOutputs` is set to true.